File: alternative/medium/maximum_product_subarray.py
from typing import List

# This is a solution for a problem frequently asked in Amazon coding interviews.
def maxProduct(nums: List[int]) -> int:
    res = max(nums)
    curMin, curMax = 1, 1

    for n in nums:
        tmp = curMax * n
        curMax = max(n * curMax, n * curMin, n)
        curMin = min(tmp, n * curMin, n)
        res = max(res, curMax, curMin)
    return res


assert maxProduct([2, 3, -2, 4]) == 6
assert maxProduct([-2, 0, -1]) == 0
assert maxProduct([-2]) == -2
assert maxProduct([0, 0, -4, 0, 1, 0, 2]) == 2

File: src/medium/maximum_product_subarray.rs
#![allow(dead_code)]
pub fn max_product(nums: Vec<i32>) -> i32 {
    let mut res = *nums.iter().max().unwrap();
    let (mut cur_min, mut cur_max) = (1, 1);

    for &n in nums.iter() {
        let tmp = cur_max * n;
        cur_max = std::cmp::max(std::cmp::max(n * cur_max, n * cur_min), n);
        cur_min = std::cmp::min(std::cmp::min(tmp, n * cur_min), n);
        res = std::cmp::max(std::cmp::max(res, cur_max), cur_min);
    }

    res
}

#[test]
fn test_max_product() {
    assert_eq!(max_product(vec![2, 3, -2, 4]), 6);
    assert_eq!(max_product(vec![-2, 0, -1]), 0);
    assert_eq!(max_product(vec![-2]), -2);
}
